src/include/function/cast/functions/cast_functions.h

```diff
@@ -34,6 +37,51 @@ struct CastRelToString {
     }
 };
 
+struct CastToUnion {
+    template<typename T>
+    static inline void operation(T&, common::union_entry_t&, common::ValueVector& inputVector,
+        common::ValueVector& resultVector) {
+        using namespace common;
+        SelectionVector* selVector = inputVector.getSelVectorPtr();
+        const auto& sourceType = inputVector.dataType;
+        const auto& targetType = resultVector.dataType;
+        uint32_t minCastCost = UNDEFINED_CAST_COST;
+        union_field_idx_t minCostField = 0;
+        for (uint64_t i = 0; i < UnionType::getNumFields(targetType); ++i) {
+            const LogicalType& fieldType = UnionType::getFieldType(targetType, i);
+            if (CastFunction::hasImplicitCast(sourceType, fieldType)) {
```

Comment:
Instead of computing the cast cost dynamically, you can know which union field you will cast to during the compile time.

Reply:
Done with https://github.com/kuzudb/kuzu/pull/5607/commits/62f10dd9949c65809452a67a93c9017efdbbb2ec